training/src/main/scala/com/airbnb/aerosolve/training/KernelTrainer.scala

```diff
@@ -0,0 +1,137 @@
+
+    val examples = LinearRankerUtils
+        .makePointwiseFloat(input, config, key)
+    val model = initModel(modelConfig, examples)
+        
+    val loss : String = modelConfig.getString("loss")
+        
+    val candidates = examples.take(candidateSize)
+
+    // Super simple SGD trainer. Mostly to get the unit test to pass
+    for (candidate <- candidates) {
+      val gradient = computeGradient(model, candidate.example(0), loss, rankKey, rankThreshold)
+      if (gradient != 0.0) {
+        val flatFeatures = Util.flattenFeature(candidate.example(0));
+        val vec = model.getDictionary().makeVectorFromSparseFloats(flatFeatures);
+        addNewSupportVector(model, kernel, scale, vec, maxSV)
+        model.onlineUpdate(gradient, learningRate, flatFeatures)
+      }
+    }
+   
+    model
+  }
+
+  def addNewSupportVector(model : KernelModel, kernel : String, scale : Float, vec : FloatVector, maxSV : Int) = {
+    val supportVectors = model.getSupportVectors()
+    if (supportVectors.size() < maxSV) {
+      val form = kernel match {
+        case "rbf" => FunctionForm.RADIAL_BASIS_FUNCTION
+        case "acos" => FunctionForm.ARC_COSINE
+        case "random" => scala.util.Random.nextInt(2) match {
+          case 0 => FunctionForm.RADIAL_BASIS_FUNCTION
+          case 1 => FunctionForm.ARC_COSINE
+        }
+      }
+      val sv = new SupportVector(vec, form, scale, 0.0f);
+      supportVectors.add(sv);
+    }
+  }
+
+  def computeGradient(model : KernelModel,
+                      fv : FeatureVector,
+                      loss : String, rankKey : String, rankThreshold : Double) : Float = {
+    val prediction = model.scoreItem(fv)
+    val label = if (loss == "hinge") TrainingUtils.getLabel(fv, rankKey, rankThreshold) else TrainingUtils.getLabel(fv, rankKey)
+    loss match {
+      case "hinge" => {
+        val lossVal = scala.math.max(0.0, 1.0 - label * prediction)
+        if (lossVal > 0.0) {
+           return -label.toFloat
+        }
+      }
+      case "regression" => {
+        val diff = prediction - label
+        if (diff > 1.0) {
+          return 1.0f
+        }
+        if (diff < -1.0) {
+          return -1.0f
+        }
+      }
+    }
+    return 0.0f;
+  }
+
+  def initModel(modelConfig : Config, examples : RDD[Example]) : KernelModel = {
+    val minCount : Int = modelConfig.getInt("min_count")
+    val rankKey : String = modelConfig.getString("rank_key")
+    
+    log.info("Building dictionary")
+    val dictionary = new StringDictionary();
+    val stats = TrainingUtils.getFeatureStatistics(minCount, examples)
+    log.info(s"Dictionary size is ${stats.size}")
+    
+    for (stat <- stats) {
+      val (family, feature) = stat._1
+      if (family != rankKey) {
+        val mean = stat._2.mean
+        val variance = if (stat._2.variance < 1e-6) 1.0 else stat._2.variance
+        var scale = Math.sqrt(1.0 / variance)
+        dictionary.possiblyAdd(family, feature, mean, scale)
```

Comment:
Not sure if I am missing anything. But for a string feature, if it's missing, featureStatistics won't see it, therefore the `mean` here would be 1, then inside `makeVectorFromSparseFloats`, the z-transform `(float) entry.scale * (feat.getValue().floatValue() - (float) entry.mean);` would transform the value of the String feature to 0.0, since for String feature its value = mean. And then when we see an example who doesn't contain this String feature, by default its value would also be 0.0, since the `FloatVector` constructor would initialize a all-zero vector. 

Then the question is how do we discriminate a missing String feature and a String feature that exists?